scarlet/source.py

```diff
@@ -635,103 +230,100 @@ def __init__(self, center, img, bg_rms, constraints=None, psf=None, symmetric=Tr
-            are contained in `morph`.
+        center: array-like
+            (y,x) coordinates of the component in the larger image
+        detect: `~numpy.array` (Ny, Nx)
         bg_cutoff: float
             Minimum non-zero flux value allowed before truncating the morphology
         """
-        # check if source_slice is covering the whole of morph:
-        # if not, extend morph by coping last row/column from img
-        if source_slice[1].stop - source_slice[1].start < self.Ny:
-            morph[0:source_slice[1].start,:] = morph[source_slice[1].start,:]
-            morph[source_slice[1].stop:,:] = morph[source_slice[1].stop-1,:]
-        if source_slice[2].stop - source_slice[2].start < self.Nx:
-            morph[:,0:source_slice[2].start] = morph[:,source_slice[2].start][:,None]
-            morph[:,source_slice[2].stop:] = morph[:,source_slice[2].stop-1][:,None]
+        # take morph from detect, same frame but shifted to center
+        Ny, Nx = detect.shape
+        if Ny % 2 == 0:
+            Ny += 1
+        if Nx % 2 ==0:
+            Nx += 1
+        morph = np.zeros((Ny,Nx))
+        im_slice, morph_slice = Component.get_frame(detect.shape, center, (Ny,Nx))
+        morph[morph_slice] = detect[im_slice]
+
+        # check if morph_slice is covering the whole of morph:
+        # if not, extend morph by coping last row/column from detect
+        if morph_slice[0].stop - morph_slice[0].start < Ny:
+            morph[0:morph_slice[0].start,:] = morph[morph_slice[0].start,:]
+            morph[morph_slice[0].stop:,:] = morph[morph_slice[0].stop-1,:]
+        if morph_slice[1].stop - morph_slice[1].start < Nx:
+            morph[:,0:morph_slice[1].start] = morph[:,morph_slice[1].start][:,None]
+            morph[:,morph_slice[1].stop:] = morph[:,morph_slice[1].stop-1][:,None]
 
         # symmetric, monotonic
         if symmetric:
             symm = np.fliplr(np.flipud(morph))
             morph = np.min([morph, symm], axis=0)
         if monotonic:
             # use finite thresh to remove flat bridges
-            from . import operators
-            prox_monotonic = operators.prox_strict_monotonic((self.Ny, self.Nx), thresh=0.1,
-                                                             use_nearest=False)
-            morph = prox_monotonic(morph, 0).reshape(self.Ny, self.Nx)
+            from . import operator
+            prox_monotonic = operator.prox_strict_monotonic((Ny, Nx), thresh=0.1, use_nearest=False)
+            morph = prox_monotonic(morph, 0).reshape(Ny, Nx)
 
         # trim morph to pixels above threshold
         mask = morph > bg_cutoff
         if mask.sum() == 0:
-            msg = "No flux above threshold={2} for source at y={0}, x={1}"
-            _y, _x = self.center_int
-            raise SourceInitError(msg.format(_y, _x, bg_cutoff))
+            msg = "No flux above threshold={1} for source at {0}"
+            raise SourceInitError(msg.format(center, bg_cutoff))
```

Comment:
I like the old message better. It explicitly stated `x=` and `y=` to try and catch people who mix up the x and y coordinates in center.

Reply:
That's fine, then it should be `y={0}, x={1}` before, which is consistent with the format of `center`. I made this more verbose after Lorena was confused about the ordering, so I think that it is good to be explicit about the `y` and `x` values here.

Reply:
I understand your point, but we've now put warnings everywhere, from the docs to the docstrings. If someone really messes this up, they won't figure it out from that error message either. The point is that the models will be all wrong, and this exception may not be thrown either.

What you'd want is an exception that says "you flipped coordinates from y/x to x/y", which requires a lot more testing that we shouldn't have to do.

Reply:
But this is the only place in the code where we explicitly say what we interpreted for `x` and `y`. I don't understand why including it is a big deal, and it can potentially save a user a lot of headaches (and save us from answering the same question over and over again).